### PR TITLE
Add trigger-contribution-build job

### PR DESCRIPTION
**Pull request by samuelFain**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
relates: [CIAC-9094](https://jira-dc.paloaltonetworks.com/browse/CIAC-9094)

## Description
Add new schedualed pipeline in GitLab to handle the "Trigger contribution build" flow as a part of this Epic: [CIAC-9092](https://jira-dc.paloaltonetworks.com/browse/CIAC-9092)

### CODE REVIEW

File: Utils/trigger_contribution_build.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+        if branch := gitlab_project.branches.get(github_branch_name):
+            print(f'--- Handling branch: {branch.name}. ---')
+
+            cancel_active_pipelines(gitlab_project, branch)
+
+            variables = {
+                "CONTRIB_BRANCH": branch.name,
+                "PULL_REQUEST_NUMBER": str(pull_request.number),
+                "CI_COMMIT_BRANCH": branch.name,
+                "CI_PIPELINE_SOURCE": "contrib",
```
Comment (JasBeilin) [inline Utils/trigger_contribution_build.py:85]: Lets change to const

File: Utils/trigger_contribution_build.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    for issue in github_issues:
+        issue.create_comment(COMMENT_MESSAGES.build_request_accepted)
+        pull_request = issue.as_pull_request()
+        github_branch_name = pull_request.head.ref
+
+        # get the corresponding GitLab branch object corresponding to the GitHub branch
+        if branch := gitlab_project.branches.get(github_branch_name):
+            print(f'--- Handling branch: {branch.name}. ---')
+
+            cancel_active_pipelines(gitlab_project, branch)
```
Comment (JasBeilin) [inline Utils/trigger_contribution_build.py:79]: Lets check if it is possible to handle errors gracefully as a comment.

**Review by JasBeilin — CHANGES_REQUESTED** [review]
Looks great!

File: .gitlab/ci/.gitlab-ci.trigger-contribution-build.yml [hunk 3]
```diff
@@ -0,0 +1,29 @@
+# This rule is to not run the build for docker update branches (for non-nightly packs)
+
+# .push-rule:
+#   rules:
+#     - !reference [.filter-non-nightly-docker-updates-rule, rules]
+#     - if: '$CI_PIPELINE_SOURCE =~ /^(push|contrib)$/'
```
[the remark this answers is not in the document]
Comment (samuelFain) [reply .gitlab/ci/.gitlab-ci.trigger-contribution-build.yml:6]: Done.

File: .gitlab/ci/.gitlab-ci.trigger-contribution-build.yml [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+  - trigger-contribution-build
+
+trigger-contribution-build:
+  stage: trigger-contribution-build
+  variables:
+    WORKFLOW: 'Trigger contribution build test'
+    GITHUB_TOKEN: $GITHUB_TOKEN
+    GITLAB_API_TOKEN: $GITLAB_API_TOKEN
+    PR_NUMBER: $PR_NUMBER
+    BASE_BRANCH: $BASE_BRANCH
```
[the remark this answers is not in the document]
Comment (samuelFain) [reply .gitlab/ci/.gitlab-ci.trigger-contribution-build.yml:22]: Done.

File: .gitlab/ci/.gitlab-ci.trigger-contribution-build.yml [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    TRIGGER_CONTRIBUTION: $TRIGGER_CONTRIBUTION
+    CONTRIB_BRANCH: $CONTRIB_BRANCH
```
[the remark this answers is not in the document]
Comment (samuelFain) [reply .gitlab/ci/.gitlab-ci.trigger-contribution-build.yml:24]: Done.

File: Utils/trigger_contribution_build.py [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+from github.Issue import Issue
+from github.PaginatedList import PaginatedList
+from gitlab import Gitlab
+from gitlab.v4.objects.branches import ProjectBranch
+from gitlab.v4.objects.projects import Project
+
+urllib3.disable_warnings()
+
+CONTRIBUTION_PRS_QUERY = 'is:pull-request state:open label:ready-for-instance-test label:Contribution label:"External PR"'
+GITLAB_PROJECT_ID = os.getenv("CI_PROJECT_ID") or 1061
```
[the remark this answers is not in the document]
Comment (samuelFain) [reply Utils/trigger_contribution_build.py:17]: Done.

File: Utils/trigger_contribution_build.py [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+MESSAGES = namedtuple(
+    "MESSAGES",
+    ["build_request_accepted", "build_triggered", "build_trigger_failed"],
+)
+COMMENT_MESSAGES = MESSAGES(
+    "For the Reviewer: Trigger build request has been accepted for this contribution PR.",
+    "For the Reviewer: Successfully created a pipeline in GitLab with url: {url}",
+    "For the Reviewer: Could not create a pipeline in GitLab for branch name: {branch}. \
+    The branch was not found in the GitLab project.",
+)
```
[the remark this answers is not in the document]
Comment (samuelFain) [reply Utils/trigger_contribution_build.py:30]: Done.

File: Utils/trigger_contribution_build.py [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+    Older active pipelines for each branch will be canceled.
+
+    Args:
+        args (Namespace): Script arguments.
+        github_issues (PaginatedList[Issue]): List of GitHub PRs returned from the contribution PRs query.
+        gitlab_project (Project): GitLab project object.
+    """
+    for issue in github_issues:
+        issue.create_comment(COMMENT_MESSAGES.build_request_accepted)
+        pull_request = issue.as_pull_request()
```
[the remark this answers is not in the document]
Comment (samuelFain) [reply Utils/trigger_contribution_build.py:72]: Done.

File: Utils/trigger_contribution_build.py (hunk 1 above)
Comment (samuelFain) [reply Utils/trigger_contribution_build.py:85]: Done.

File: Utils/trigger_contribution_build.py [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+
+    Args:
+        gitlab_project (Project): GitLab Project object.
+        branch (ProjectBranch): Gitlab branch object.
+    """
+    pipelines = gitlab_project.pipelines.list(ref=branch.name, status="running")
+    if pipelines:
+        for pipeline in pipelines:
+            print(f"Canceling active pipeline: {pipeline.id}")
+            pipeline.cancel()
```
[the remark this answers is not in the document]
Comment (samuelFain) [reply Utils/trigger_contribution_build.py]: Logs added for the entire script.

File: Utils/trigger_contribution_build.py (hunk 2 above)
Comment (samuelFain) [reply Utils/trigger_contribution_build.py]: try-except has been added where necessary.

File: .gitlab/ci/.gitlab-ci.trigger-contribution-build.yml [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+trigger-contribution-build:
+  stage: trigger-contribution-build
+  variables:
+    WORKFLOW: 'Trigger contribution build'
+    GITHUB_TOKEN: $GITHUB_TOKEN
+    GITLAB_API_TOKEN: $GITLAB_API_TOKEN
+    TRIGGER_CONTRIBUTION: $TRIGGER_CONTRIBUTION
+  extends:
+    - .trigger-contribution-build-role
+    - .default-job-settings
```
Comment (michal-dagan) [inline .gitlab/ci/.gitlab-ci.trigger-contribution-build.yml:17]: Are all the .default-job-settings needed?

File: Utils/trigger_contribution_build.py [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+from gitlab import Gitlab
+from gitlab.exceptions import GitlabError
+from gitlab.v4.objects.branches import ProjectBranch
+from gitlab.v4.objects.projects import Project
+from Tests.scripts.utils.log_util import install_logging
+
+urllib3.disable_warnings()
+
+DEFAULT_CI_PIPELINE_SOURCE = "contrib"
+GITLAB_DEFAULT_PROJECT_ID = 1061
```
Comment (michal-dagan) [inline Utils/trigger_contribution_build.py:19]:
```suggestion
GITLAB_DEFAULT_PROJECT_ID = os.getenv('CI_PROJECT_ID') or 1061
```

File: Utils/trigger_contribution_build.py [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+CONTRIBUTION_PRS_QUERY = 'is:pull-request state:open label:ready-for-instance-test label:Contribution label:"External PR"'
+GITLAB_PROJECT_ID = os.getenv("CI_PROJECT_ID") or GITLAB_DEFAULT_PROJECT_ID
```
Comment (michal-dagan) [inline Utils/trigger_contribution_build.py:21]: maybe merge it to the same variable: GITLAB_DEFAULT_PROJECT_ID?

**Review by michal-dagan — APPROVED** [review]
LGTM, Great work!

File: .gitlab/ci/.gitlab-ci.trigger-contribution-build.yml (hunk 10 above)
Comment (samuelFain) [reply .gitlab/ci/.gitlab-ci.trigger-contribution-build.yml:17]: Not sure, but It is needed for the logs mechanism to work.

File: Utils/trigger_contribution_build.py (hunk 11 above)
Comment (JasBeilin) [reply Utils/trigger_contribution_build.py:19]: We dont want the constant to be define by the env vars, I think we should keep this a const..

**Review by JasBeilin — APPROVED** [review]
Looks Good.

**Comment by samuelFain** [comment]
Tested in Nightly pipeline: https://gitlab.xdr.pan.local/xdr/cortex-content/content/-/pipelines/923946

**Pull request merged**